src/components/js-tracer/Chart.js

```diff
@@ -0,0 +1,260 @@
+  wasLoaderMounted: boolean,
+|};
+
+// Keep track of all the React keys seen for a component. If everything is correctly
+// memoized, then it should only be slow and expensive to compute the timing information
+// the first time. After that we don't need to use a loader/fading in strategy.
+//
+// These sets are valid per profile, so WeakMap them on the profile.
+const _seenChartKeysPerProfile: WeakMap<Profile, Set<string>> = new WeakMap();
+
+/**
+ * This component displays a helpful loading screen for the first time a selector is
+ * run. It relies on having a propery `key` property on the component. This way
+ * initialization and invalidation is all handled through the component lifecycles.
+ */
+class JsTracerChartLoader extends React.PureComponent<
+  ChartLoaderProps,
+  ChartLoaderState
+> {
+  state = { wasLoaderMounted: false };
+  _doFadeIn: boolean = false;
+
+  constructor(props) {
+    super(props);
+    // Look up the seenChartKeys per-profile. If not found, create a new Set.
+    let seenChartKeys = _seenChartKeysPerProfile.get(props.profile);
+    if (seenChartKeys === undefined) {
+      seenChartKeys = new Set();
+      _seenChartKeysPerProfile.set(props.profile, seenChartKeys);
+    }
+
+    if (!seenChartKeys.has(props.keyString)) {
+      // Only fade in the component on the first render. After the first time,
+      // everything should be correctly memoized, and not need another re-render.
+      this._doFadeIn = true;
+      seenChartKeys.add(props.keyString);
+    }
+  }
+
+  componentDidMount() {
+    if (this._doFadeIn) {
+      // Let the screen render at least once, then start computing the expensive chart.
+      requestAnimationFrame(() => {
+        requestAnimationFrame(() => {
+          this.setState({ wasLoaderMounted: true });
+        });
+      });
+    }
+  }
+
+  render() {
+    const { jsTracerTable, showJsTracerSummary } = this.props;
+    return this.state.wasLoaderMounted || !this._doFadeIn ? (
+      <JsTracerExpensiveChart
+        doFadeIn={this._doFadeIn}
+        jsTracerTable={jsTracerTable}
+        showJsTracerSummary={showJsTracerSummary}
+      />
+    ) : (
+      <div className="jsTracerLoader">
+        Re-constructing tracing information from{' '}
+        {this.props.jsTracerTable.events.length.toLocaleString()} events. This
+        might take a moment.
+      </div>
+    );
+  }
+}
+
+type ChartProps = {|
+  +profile: Profile,
+  +jsTracerTable: JsTracerTable,
+  +showJsTracerSummary: boolean,
+  +threadIndex: ThreadIndex,
+|};
+
+/**
+ * This component enforces that the JsTracerChartLoader has a correct key in order
+ * to signal React to unmount the previous JsTracerChartLoader componenent, and create
+ * a new one. This ensures that expensive selectors won't re-run as we are following
+ * the React component lifecycle.
+ *
+ * See: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key
```

Comment:
The react website is broken: the link doesn't move to the right paragraph :( (not sure what we can do about that though)